### ISSUE TITLE
AttributeError: 'NoneType' object has no attribute 'shape'

### ISSUE THREAD

**Issue by kommandant-gm** [FaceGg/Face-Anti-Spoofing#1]
Hi when i try to run this error comes out : 
AttributeError: 'NoneType' object has no attribute 'shape'

**Comment by FaceGg** [comment]
> Hi when i try to run this error comes out :
> AttributeError: 'NoneType' object has no attribute 'shape'

@kommandant-gm  you need have a video named  '1.mp4' . at line 90 of the 'demo.py',  `vs = cv2.VideoCapture('1.mp4')`,  we use opencv to open a video, you can change the '1.mp4' into name of your video

**Comment by kommandant-gm** [comment]
So basically i can use my own video dataset and change the named to 1.mp4 ? am i correct ? Do u might have any reference ?

**Comment by FaceGg** [comment]
@kommandant-gm 
Yes, this video is to test the effect, not the training data, so you can call it any name

**Comment by kommandant-gm** [comment]
can i follow ur method using opencv as well ? or do i just need to record the face and named it 1.mp4 ?

**Comment by FaceGg** [comment]
@kommandant-gm 
You can take a video yourself or download one from the Internet.If your computer has a camera, you can also test the real-time video.If there is one or more  face in the video, the program will first detect the face, and then judge is real or fake. if there is no face, program will do nothing.

**Comment by kommandant-gm** [comment]
Thanks for your assist. I understand about the video problem where i can record myself but may i ask one more how do i get real time video to work ? any specific steps that i have to do ? I do have camera as im using laptop. just the problem when i try to run the file the error pop out

**Comment by FaceGg** [comment]
@kommandant-gm 
it is easy.
'demo.py'
line 89   #vs = VideoStream(src=0).start()        
line 90   vs = cv2.VideoCapture('1.mp4')
line 94   #frame = vs.read()       
line 95   frame = vs.read()[1]
change them into 
line 89   vs = VideoStream(src=0).start()         
line 90   #vs = cv2.VideoCapture('1.mp4')
line 94   frame = vs.read()           
line 95   #frame = vs.read()[1]

**Comment by kommandant-gm** [comment]
I have changed the step accordingly but these is the error that pop out

https://drive.google.com/open?id=1ZIcLjt4qPB_Y68jbn5gCvOu-OoSF1PJ1

**Comment by FaceGg** [comment]
@kommandant-gm 
This means the program can't find the camera. You can check if your camera driver is installed correctly or not , and make sure the program has permission to turn on your camera

**Comment by kommandant-gm** [comment]
i tried previously doing the face allignment detection project that u post as well and it works just fine. it can open up the camera on my laptop and detect face. 

But now i cant open up for this project. Is there any setting that i need configure ?

**Comment by FaceGg** [comment]
@kommandant-gm i just run it , have no problem. you can show me the codes form line 89 to line 100

**Comment by kommandant-gm** [comment]
yes this the new error pop out 

Traceback (most recent call last):
  File "C:/Users/User/Desktop/Face-anti-spoofing-master/demo.py", line 102, in <module>
    detections = net.forward()
cv2.error: OpenCV(4.1.1) C:\projects\opencv-python\opencv\modules\dnn\src\layers\convolution_layer.cpp:306: error: (-2:Unspecified error) Number of input channels should be multiple of 3 but got 1 in function 'cv::dnn::ConvolutionLayerImpl::getMemoryShapes'

Code from like 89 to 105 : 

print("[INFO] starting video stream...")
    vs = VideoStream(src=0).start() #vs = cv2.VideoCapture('1.mp4')
  
    time.sleep(2.0)

    while True:
        frame = vs.read()  # frame = vs.read()[1]
        frame = vs.read()[1]
        frame = imutils.resize(frame, width=600)
        #frame = imutils.rotate(frame, -90)
        (h, w) = frame.shape[:2]
        blob = cv2.dnn.blobFromImage(cv2.resize(frame, (300, 300)), 1.0,(300, 300), (104.0, 177.0, 123.0))
        net.setInput(blob)
        start = time.time()
        detections = net.forward()
        end = time.time()

        print('detect times : %.3f ms'%((end - start)*1000))

**Comment by FaceGg** [comment]
@kommandant-gm 
ok ,I know what the problem is. i have changed the codes of 'demo.py', you can try the new

**Comment by kommandant-gm** [comment]
Hi it works now. but it show my face is fake. how do i train own dataset ? 
Apparently if my face arent that close to camera it will detect as fake

**Comment by FaceGg** [comment]
@kommandant-gm 
You can try to get close to the camera, if the face is too small ,it will think it is fake .  also you can change the thresholds at line 86.  About the training codes ,i am sorry about that it will not be open-sourcing now.

**Comment by kommandant-gm** [comment]
I see appreciate much of your help but may i know does these project use CNN for training dataset ?

**Comment by FaceGg** [comment]
@kommandant-gm 
yes , using CNN is an inevitable trend.

**Comment by kommandant-gm** [comment]
Hi, forgot to ask you something about the project i noticed that in the frame beside the true/fake there are numbers. Is it represent the accuracy ?

**Comment by FaceGg** [comment]
@kommandant-gm 
This is a binary classification, the outputs of model is that output[0] is the probability of fake face , the output[1] is the probability of real face , We focus on the probability of fake face, so we judge if output[0] > thresh( line 86 ), we think that it is a fake face .

**Comment by kommandant-gm** [comment]
Hi i have question again sorry. How does the program actually detect it as real or fake ? Is it liveliness detection ?
